### PR TITLE
Automate release preparation process

### PR DESCRIPTION
**Pull request by ClaudiaComito**
## Due Diligence

- General:
    - [x]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [ ] unit tests: all split configurations tested
    - [ ] unit tests: multiple dtypes tested
    - [ ] benchmarks: created for new functionality
    - [ ] benchmarks: performance improved or maintained
    - [x] documentation updated where needed

## Summary

This PR introduces a new GitHub Actions workflow named "Prep Heat Release". The workflow is designed to automate the process of preparing a release for the Heat project.

## Motivation

The goal of this workflow is to streamline the release preparation process by automating version extraction, branch creation, and version bumping. This reduces manual effort and minimizes the risk of human error.

## Workflow Details

The workflow is triggered when an issue is opened with the label `release-prep`. A dedicated issue template has been added to the repo.

## Impact

These changes will automate the release preparation process. It will create new release branches if needed, and update version numbers automatically based on the issue title. It will create a PR where further (manual) work can be collected.

The resulting PR looks like [this](https://github.com/ClaudiaComito/heat/pull/31).

Issue/s resolved: #1709 

## Changes proposed:

-  Added a new issue template for release preparation in `.github/ISSUE_TEMPLATE/release_prep.md` to guide users through the process of triggering the release preparation workflow.
- Created a new GitHub Actions workflow in `.github/workflows/release-prep.yaml` to automate the release preparation process, including extracting version information, creating release branches, updating version numbers, and creating pull requests.
- Updated `RELEASE.md` to reflect the new automated release preparation process, including instructions on how to trigger the workflow, update version numbers, and publish releases. [[1]](diffhunk://#diff-2b1b69303b927a484e02c7fad9fc87d0d3ff0dc22ae1da0ecd0dc935d922a23cL6-R31) [[2]](diffhunk://#diff-2b1b69303b927a484e02c7fad9fc87d0d3ff0dc22ae1da0ecd0dc935d922a23cL83-R59) [[3]](diffhunk://#diff-2b1b69303b927a484e02c7fad9fc87d0d3ff0dc22ae1da0ecd0dc935d922a23cL113-R72)
- Added placeholders for additional release methods such as easybuild, spack, and Docker in `RELEASE.md`.
-
-
-

## Type of change

Automation   and docs update
#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

File: .github/workflows/release-prep.yaml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+          echo "major=$major" >> $GITHUB_OUTPUT
+          echo "minor=$minor" >> $GITHUB_OUTPUT
+          echo "micro=$micro" >> $GITHUB_OUTPUT
+      - name: Delete Existing Version Update Branch
+        run: git push --delete origin ${{ env.working_branch }} || true
+      - name: Create Release Branch
+        if: steps.extract-version.outputs.micro == 0
+        run: |
+          git checkout -b release/${{ steps.extract-version.outputs.major }}.${{ steps.extract-version.outputs.minor }}.x
+          git push origin release/${{ steps.extract-version.outputs.major }}.${{ steps.extract-version.outputs.minor }}.x
```
Comment (JuanPedroGHM) [inline .github/workflows/release-prep.yaml:40]: In keeping with git-flow style branches, I would call this "pre-release/x.y.z". Based on [this link](https://jeffkreeftmeijer.com/git-flow/), "main" is would be "release" in our case, "develop" is "main", and we could name "release/x.y.z" as "pre-release/x.y.z".

File: .github/workflows/release-prep.yaml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        run: |
+          issue_title="${{ github.event.issue.title }}"
+          major=$(echo "$issue_title" | cut -d '.' -f 1)
+          minor=$(echo "$issue_title" | cut -d '.' -f 2)
+          micro=$(echo "$issue_title" | cut -d '.' -f 3)
+          echo "major=$major" >> $GITHUB_OUTPUT
+          echo "minor=$minor" >> $GITHUB_OUTPUT
+          echo "micro=$micro" >> $GITHUB_OUTPUT
+      - name: Delete Existing Version Update Branch
+        run: git push --delete origin ${{ env.working_branch }} || true
```
Comment (JuanPedroGHM) [inline .github/workflows/release-prep.yaml:35]: When is this branch created, or where does it come from?

File: .github/workflows/release-prep.yaml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+          git push origin ${{ env.working_branch }}
+          sed -i "s/major: int = \([0-9]\+\)/major: int = ${{ steps.extract-version.outputs.major }}/g" heat/core/version.py
+          sed -i "s/minor: int = \([0-9]\+\)/minor: int = ${{ steps.extract-version.outputs.minor }}/g" heat/core/version.py
+          sed -i "s/micro: int = \([0-9]\+\)/micro: int = ${{ steps.extract-version.outputs.micro }}/g" heat/core/version.py
+          sed -i "s/extension: str = .*/extension: str = None/g" heat/core/version.py
+      - name: Create PR from branch
+        uses: peter-evans/create-pull-request@5e914681df9dc83aa4e4905692ca88beb2f9e91f # v7.0.5
+        with:
+            base: release/${{ steps.extract-version.outputs.major }}.${{ steps.extract-version.outputs.minor }}.x
+            branch: ${{ env.working_branch }}
```
Comment (JuanPedroGHM) [inline .github/workflows/release-prep.yaml:56]: The base branch should be "release", and the branch should be our newly created branch, right? "pre-release/x.y.z"

File: .github/workflows/release-prep.yaml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+          micro=$(echo "$issue_title" | cut -d '.' -f 3)
+          echo "major=$major" >> $GITHUB_OUTPUT
+          echo "minor=$minor" >> $GITHUB_OUTPUT
+          echo "micro=$micro" >> $GITHUB_OUTPUT
+      - name: Delete Existing Version Update Branch
+        run: git push --delete origin ${{ env.working_branch }} || true
+      - name: Create Release Branch
+        if: steps.extract-version.outputs.micro == 0
+        run: |
+          git checkout -b release/${{ steps.extract-version.outputs.major }}.${{ steps.extract-version.outputs.minor }}.x
```
Comment (JuanPedroGHM) [inline .github/workflows/release-prep.yaml:39]: I would add an extra `git checkout main; git pull` to make sure we are branching of the most recent version of main.

**Review by JuanPedroGHM — COMMENTED** [review]
Missing:
- Name adjustment to new branching style.

Ideas:
- To PR's could be created, the one that will be merged into release, and the one that will be merged into main. Both should follow the newly created branch "pre-release/x.y.z"

**Comment by ClaudiaComito** [comment]
Closing as this will be addressed in #1678

**Pull request closed without merging**